### DISCUSSION TITLE
Classmethod usage inside a table model

### DISCUSSION THREAD

**Discussion by dakide** [sqlalchemy/sqlalchemy#9446] [Usage Questions] [answered]
This is a migration sort of question. 
I'm pretty new to SQLAlchemy, I've looked through the documentation and migration docs but I can't find how to accomplish what I want.

I'm used to adding classmethods to my models, for example: 

```python

class Kudos(db.Model):
    __tablename__ = "kudos"

    id = Column(Integer, primary_key=True)
    given_on = Column(DateTime)
    given_by = Column(String(50))
    deed = Column(String(1000))
    active = Column(Boolean)
    
    @classmethod
    def find_all(cls) -> "Kudos":
        return cls.query.filter_by(active=True)

```

How would that look in SQLA 2.0?

If I add that classmethod to the new 2.0 model (using mapped_column etc) the program fails because the Kudos object doesn't have a "query" attribute
`AttributeError: type object 'Kudos' has no attribute 'query'`

**Comment by zzzeek** [comment, marked as answer]
SQLAlchemy by itself doesn't add that .query attribute, that's a flask-sqlalchemy thing.    I would have thought they'd kept that in place, since that feature still works in SQLAlchemy 2.0.

your find_all() method seems to return a new Query object so you would return a select() object instead:

```py
from sqlalchemy import select

# ...

    @classmethod
    def find_all(cls) -> "Kudos":
        return select(cls).filter_by(active=True)

```

your application would then have `Session` available for all database methods and you'd use that to `.execute()`.

**Reply by dakide** [reply]
Thanks for that.

I've changed the model to this:
```python
 from sqlalchemy.orm import Session

    ...

    @classmethod
    def find_all(cls, session: Session):
        return session.execute(
            select(cls).where(
                cls.active == True
            )
        ).scalar_one_or_none()

```

Then in the main application I am using this:
```python
  kudos = Kudos.find_all(session=db_session)
```

Is that an acceptable/efficient/correct way of doing it?

**Reply by zzzeek** [reply]
yeah that's a great way!

**Reply by CaselIT** [reply]
that should work. mostly. note that you probably don't want to call `scalar_one_or_none` there, since it will raise an error if more than a row is matched.

probably using `session.scalars(...).all()` is a better fit for a `find_all` method

**Reply by dakide** [reply]
Great, thanks for your help!
